### DISCUSSION TITLE
Expose Supabase-Admin-API to Kong

### DISCUSSION THREAD

**Discussion by Garett-MacGowan** [supabase/supabase#6682] [Feature Requests]
# Feature request

## Problem
It is ~~currently impossible~~ **see comment below for workaround** to programmatically add schemas to the API/Realtime. The only way to add schemas as of this request is to navigate to `settings/api` and toggle the desired schema in the form.

![image](https://user-images.githubusercontent.com/24864607/166126260-d6608b4e-e09d-4b61-9a5c-a00cd8787538.png)

Many users might want to programmatically add schemas to the API if they frequently add new schemas for new feature sets.

It is also very difficult to manually add schemas to a [local instance](https://supabase.com/docs/guides/local-development) since `settings/api` isn't exposed in the local studio.

![unknown](https://user-images.githubusercontent.com/24864607/166126167-78746a78-2d9a-4d03-a76f-3388c8573b67.png)

## Potential solution

I think the best solution would be to expose [Supabase Admin API](https://github.com/supabase/supabase-admin-api) to Kong so that administrators can update configs programmatically.

![Untitled](https://user-images.githubusercontent.com/24864607/166126327-ddda0fde-5a10-4f8a-a281-fed3013e2d18.png)

## Alternatives

I'm ~~currently~~ **UPDATE see comment below for my new strategy (which works fine, but is slightly complicated)** manually adding schemas through the `settings/api` page of the studio. Although I can't do this for local development, so I'm running E2E tests through a hosted dev Supabase instance for now. This is dangerous because I could have contention between developers and/or test runners.

## Useful docs

[PostgREST configuration](https://postgrest.org/en/stable/configuration.html)

**Comment by Garett-MacGowan** [comment]
Thanks to an example sent to me by Steve in the discord, I have managed to get db_schema updates working through an RPC call with Kong. I still think the feature request is valid, since the solution is a little messy, but I'd be glad to hear others thoughts.

Here's a quick summary:
* Create a postgres function for changing the authenticator role's pgrst.db_schemas attribute, ensuring to keep the defaults (public, storage). E.g.
```
CREATE FUNCTION change_postgrest_db_schemas(schemas text) RETURNS VOID AS $_$
BEGIN
    EXECUTE format($$ 
        ALTER ROLE authenticator SET pgrst.db_schemas = %L; 
    $$, schemas); 
    perform pg_notify('pgrst', 'reload config'); 
    perform pg_notify('pgrst', 'reload schema');
END $_$ volatile security definer language plpgsql;
```
Above code is based on a [link](https://github.com/PostgREST/postgrest/blob/main/test/io/fixtures.sql#L49-L56) provided by Steve.

* Trigger the function with a PostgREST rpc call (via Kong)
  * Be sure to always include defaults here (public, storage) 
```
POST http://localhost:54321/rest/v1/rpc/change_postgrest_db_schemas
payload {"schemas": "public, storage, new_schema"}
headers apikey=service_key
```

In order to make my solution above robust, I ended up creating another postgres function that checks what the current state of pgrst.db_schemas is, so that I can update the current state.

**Comment by MildTomato** [comment]
Moving this issue to 'idea' discussion board

**Comment by nmabhinandan** [comment]
This essentially makes it impossible to develop the projects locally who are using custom schema. Until this feature is added, it will be helpful to add the above workaround to the documentation here.

https://supabase.com/docs/guides/api/using-custom-schemas
